File: python/pyduct/core/geometry.py
"""Cross-section geometry primitives.

A `CrossSection` is an immutable value object that knows its area and
hydraulic diameter. Both are computed once in ``__post_init__`` and cached
as instance attributes, so per-call sizing loops avoid repeated `math`.
"""

from __future__ import annotations

from dataclasses import dataclass
from math import pi


class CrossSection:
    """Base class for duct cross-sections."""

    area: float
    hydraulic_diameter: float


@dataclass(frozen=True)
class Round(CrossSection):
    """Circular cross-section."""

    diameter: float  # [m]

    def __post_init__(self) -> None:
        if self.diameter <= 0:
            raise ValueError(f"diameter must be positive, got {self.diameter}")
        object.__setattr__(self, "area", pi * (self.diameter / 2) ** 2)
        object.__setattr__(self, "hydraulic_diameter", self.diameter)


@dataclass(frozen=True)
class Rectangular(CrossSection):
    """Rectangular cross-section."""

    width: float   # [m]
    height: float  # [m]

    def __post_init__(self) -> None:
        if self.width <= 0 or self.height <= 0:
            raise ValueError(
                f"width and height must be positive, got "
                f"width={self.width}, height={self.height}"
            )
        object.__setattr__(self, "area", self.width * self.height)
        # D_h = 4 A / P = 2 W H / (W + H)
        object.__setattr__(
            self,
            "hydraulic_diameter",
            2 * self.width * self.height / (self.width + self.height),
        )


def equivalent_round_diameter(width: float, height: float) -> float:
    """ASHRAE equivalent round diameter for a rectangular duct.

    ``D_eq = 1.30 · (a·b)^0.625 / (a + b)^0.25`` [m]. Math runs in
    ``wenta.core.geometry.equivalent_round_diameter``.
    """
    from wenta.ext.leaf_ext import equivalent_round_diameter as _eq
    try:
        return _eq(width, height)
    except Exception as e:
        raise ValueError(str(e)) from e

File: csharp/Wenta.Core/Geometry.cs
using System;

namespace Wenta
{
    /// <summary>Cross-section geometry primitives. Port of `wenta.core.geometry`.
    /// Area and hydraulic diameter are computed once at construction.</summary>
    public abstract class CrossSection
    {
        public readonly double Area;             // [m^2]
        public readonly double HydraulicDiameter; // D_h [m]
        public abstract string Describe();

        protected CrossSection(double area, double hydraulicDiameter)
        {
            Area = area;
            HydraulicDiameter = hydraulicDiameter;
        }
    }

    /// <summary>Circular cross-section.</summary>
    public sealed class Round : CrossSection
    {
        public readonly double Diameter; // [m]

        public Round(double diameter)
            : base(Math.PI * (diameter / 2.0) * (diameter / 2.0), diameter)
        {
            if (diameter <= 0.0)
                throw new WentaException("diameter must be positive, got " + diameter);
            Diameter = diameter;
        }

        public override string Describe() { return "round D=" + (Diameter * 1000) + "mm"; }
    }

    /// <summary>Rectangular cross-section.</summary>
    public sealed class Rectangular : CrossSection
    {
        public readonly double Width;  // [m]
        public readonly double Height; // [m]

        public Rectangular(double width, double height)
            : base(width * height, 2.0 * width * height / (width + height))
        {
            if (width <= 0.0 || height <= 0.0)
                throw new WentaException(
                    "width and height must be positive, got width=" + width
                    + ", height=" + height);
            Width = width;
            Height = height;
        }

        public override string Describe()
        {
            return "rect " + (Width * 1000) + "x" + (Height * 1000) + "mm";
        }
    }

    public static class Geometry
    {
        /// <summary>ASHRAE equivalent round diameter for a rectangular duct:
        /// D_eq = 1.30 · (a·b)^0.625 / (a + b)^0.25 [m].</summary>
        public static double EquivalentRoundDiameter(double width, double height)
        {
            if (width <= 0.0 || height <= 0.0)
                throw new WentaException("width and height must be positive");
            return 1.30 * Math.Pow(width * height, 0.625) / Math.Pow(width + height, 0.25);
        }
    }
}
